### PR TITLE
Better Prefixes

### PR DESCRIPTION
**Pull request by A5rocks**
Make bots be able to have multiple prefixes, and maybe even eventually having programmatic prefix fetching to allow for custom prefixes. Why not make the library user do this work? Because it's common enough and not everyone wants to copy and paste the source into a handler just to edit a couple lines.

Example:
image

### CODE REVIEW

File: disco/bot/bot.py [hunk 1]
```diff
@@ -275,21 +279,33 @@ def compute_command_matches_re(self):
...
         Generator of all commands that a given message object triggers, based on
         the bots plugins and configuration.
 
         Parameters
         ---------
+        require_mention : bool
+            Checks if the message starts with a mention (and then ignores the prefix(es))
+        mention_rules : dict(str, bool)
+            Whether `user`, `everyone`, and `role` mentions are allowed. Defaults to:
+            `{'user': True, 'everyone': False, 'role': False}
```
Comment (A5rocks) [inline disco/bot/bot.py]: An ending "`" is missing, but that's not enough to warrant an extra commit.
```suggestion
            `{'user': True, 'everyone': False, 'role': False}`
```

**Comment by A5rocks** [comment]
custom prefixes get!

File: disco/util/runner.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    Any = str  # just give up on typing...
+
+# imports from disco (if moved to the top, they will probably break
+# due to requiring `gevent`'s monkey patching for asynchronous)
+from disco.client import Client, ClientConfig       # noqa: E402
+from disco.bot import Bot, BotConfig                # noqa: E402
+from disco.util.logging import setup_logging        # noqa: E402
+
+
+def bot_creator(config=None, bot=True, autosharded=False, **kwargs):
```
Comment (b1naryth1ef) [inline disco/util/runner.py]: This should be moved into a sep. PR if you'd like it included.

File: disco/bot/bot.py [hunk 3]
```diff
@@ -46,6 +46,12 @@ class BotConfig(Config):
     commands_prefix : str
         A string prefix that is required for a message to be considered for
         command parsing.
+    commands_prefixes : list[string]
```
Comment (b1naryth1ef) [inline disco/bot/bot.py]: If we're adding this configuration option, we need to deprecate `commands_prefix`. This probably means we need to throw a DeprecationWarning when we read/validate the config inside of the bot constructor.

File: disco/bot/bot.py [hunk 4]
```diff
@@ -326,11 +346,18 @@ def get_commands_for_message(self, require_mention, mention_rules, prefix, msg):
                     content = content.replace('<@{}>'.format(role), '', 1)
 
             content = content.lstrip()
-
-        if prefix and not content.startswith(prefix):
-            return []
         else:
-            content = content[len(prefix):]
+            # Scan through the prefixes to find the first one that matches.
```
Comment (b1naryth1ef) [inline disco/bot/bot.py]: This block doesn't logically make sense to be a else branch of the prior if, they are unrelated.

File: disco/bot/bot.py [hunk 5]
```diff
@@ -326,11 +346,18 @@ def get_commands_for_message(self, require_mention, mention_rules, prefix, msg):
...
+            # This may lead to unexpected results, but said unexpectedness
+            # should be easy to avoid. An example of the unexpected results
+            # that may occur would be if one prefix was `!` and one was `!a`.
+            if not any([content.startswith(prefix) for prefix in prefixes]):
```
Comment (b1naryth1ef) [inline disco/bot/bot.py]: This code shouldn't check over the message content twice, just do the following;

```python
for prefix in prefixes:
    if content.startswith(prefix):
        content = content[len(prefix):]
        break
else:
    return []
```

It may also make sense to give the end user access to the prefix that was matched in the command object somewhere.

File: disco/bot/bot.py [hunk 6]
```diff
@@ -383,11 +411,17 @@ def handle_message(self, msg):
         bool
             Whether any commands where successfully triggered by the message.
         """
+        if not self.config.commands_require_mention and self.config.commands_prefix_getter:
```
Comment (b1naryth1ef) [inline disco/bot/bot.py]: This should just be;

```python
custom_message_prefixes = self.config.commands_prefix_getter(msg) if self.config.commands_prefix_getter is not None
```

File: disco/bot/bot.py [hunk 7]
```diff
@@ -383,11 +411,17 @@ def handle_message(self, msg):
...
+            custom_message_prefixes = self.config.commands_prefix_getter(msg)
+        else:
+            custom_message_prefixes = []
+
         commands = list(self.get_commands_for_message(
             self.config.commands_require_mention,
             self.config.commands_mention_rules,
             self.config.commands_prefix,
             msg,
+            self.config.commands_prefixes + custom_message_prefixes,
```
Comment (b1naryth1ef) [inline disco/bot/bot.py]: I'd argue it makes more sense for `custom_message_prefixes` to not be merged with the original commands_prefixes. This gives the end user more control when using the functional version of this configuration option (e.g. blocking some prefixes from matching in certain channels, or having server specific prefixes).

File: tests/bot/bot.py [hunk 8]
```diff
@@ -92,3 +92,12 @@ def test_group_and_command(self):
         msg.content = '!test'
         commands = list(self.bot.get_commands_for_message(False, None, '!', msg))
         self.assertEqual(commands[0][0], self.bot._commands[0])
+
+        msg.content = '?test a'
+        commands = list(self.bot.get_commands_for_message(False, None, '!', msg, ['?', ';']))
```
Comment (b1naryth1ef) [inline tests/bot/bot.py]: We can probably update the majority of these tests if we deprecate single-prefix.

File: disco/bot/bot.py (hunk 5 above)
Comment (A5rocks) [reply disco/bot/bot.py]: The thing is, that doesn't work. I would use an `else` after a `for` if I could though. The whole point of the overarching if statement is to tell if any prefixes will match, instead of having to use another variable or something to check if the message actually starts with a prefix.

Comment (b1naryth1ef) [reply disco/bot/bot.py]: Why doesn't it work?

Comment (A5rocks) [reply disco/bot/bot.py]: Oh wait. Oops. Today I learned...

**Comment by A5rocks** [comment]
Is there anything else I need to do?

File: disco/bot/bot.py [hunk 9]
```diff
@@ -327,10 +346,17 @@ def get_commands_for_message(self, require_mention, mention_rules, prefix, msg):
...
             content = content.lstrip()
 
-        if prefix and not content.startswith(prefix):
-            return []
+        # Scan through the prefixes to find the first one that matches.
+        # This may lead to unexpected results, but said unexpectedness
+        # should be easy to avoid. An example of the unexpected results
+        # that may occur would be if one prefix was `!` and one was `!a`.
+        for prefix in prefixes:
+            if prefix and content.startswith(prefix) and not require_mention:  # don't cut off commands with mentions.
```
Comment (b1naryth1ef) [inline disco/bot/bot.py]: I don't understand what the check to `require_mention` is doing here? We process mentions out of the message above if `reuqire_mention` is set, so this is presumably unnecessary?

**Comment by b1naryth1ef** [comment]
@A5rocks just want to understand that last piece of your changeset and then it'd be g2g

File: disco/bot/bot.py (hunk 9 above)
Comment (A5rocks) [reply disco/bot/bot.py]: I honestly don't know. Maybe I was thinking of the case where `{@mention} {prefix}` is sent when `{prefix}` is a command, but now that I think of it, this case would only be reached by people who willing configure their bots to have this behaviour (set require mention to true and have a prefix).

tl;dr I'll fix it
